### PR TITLE
Rewrite the exception handling tutorial

### PR DESCRIPTION
**Pull request by snoyberg**

### CODE REVIEW

File: content/tutorials/exceptions.md [hunk 1]
```diff
@@ -2,379 +2,697 @@
...
+
+What we won't do here:
+
+* Cover the full motivation for the design of the libraries in question
+* Debate the merits of runtime exceptions
+* Debate the merits of asynchronous exceptions
+
+## `UnliftIO.Exception`
+
+Instead of use the `Control.Exception` module, we recommend using the
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/use/using ?

File: content/tutorials/exceptions.md [hunk 2]
```diff
@@ -2,379 +2,697 @@
...
+  logInfo $ display message
+```
+
+An asynchronous exception is totally different. It is a demand from
+outside of our control to shut down as soon as possible. If we were to
+catch such an exception and recover from it, we would be breaking the
+expectations of the thread that tried to shut us down. Instead, with
+asynchronous exceptions, we're allowed to clean up, but not
+recover. For example, the `timeout` function uses asynchronous
+exceptions. What should the expected behavior here be?
```
Comment (psibi) [inline content/tutorials/exceptions.md:242]: > Instead, with asynchronous exceptions, we're allowed to clean up, but not recover. 

Is this the best practice or does the Haskell runtime itself doesn't allow us to recover ? Clarifying that would be helpful IMO.

File: content/tutorials/exceptions.md [hunk 3]
```diff
@@ -2,379 +2,697 @@
...
+
+main :: IO ()
+main = runSimpleApp $ do
+  res <- timeout 1000000 $ do
+    logInfo "Inside the timeout"
+    res <- tryAny $ threadDelay 5000000 `finally`
+      logInfo "Inside the finally"
+    logInfo $ "Result: " <> displayShow res
+  logInfo $ "After timeout: " <> displayShow res
+```
```
Comment (psibi) [inline content/tutorials/exceptions.md:263]: Probably the code can be made more easy to (lazy) readers like me by using these constants:

```
oneMinute :: Int
oneMinute = 1000000

fiveMinute :: Int
fiveMinute = 5000000
```

File: content/tutorials/exceptions.md [hunk 4]
```diff
@@ -2,379 +2,697 @@
...
+
+Bad async exception handling would allow the "Result: " message to
+print. We don't want that to happen! Instead, we allow the `finally`
+cleanup call to occur an then immediately exit. This ensures that
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/an/and ?

File: content/tutorials/exceptions.md [hunk 5]
```diff
@@ -2,379 +2,697 @@
...
+an exception how it was thrown, making it difficult to live up to our
+goals above to never recover from an async exception. Fortunately, we
+have a workaround: use a different type for async exceptions!
+
+Using the hierarchical exception mechanism above, we have a new data
+type, `SomeAsyncException`, which is a child of `SomeException`. All
+exceptions which are thrown asynchronously must be a child of that
+exception type. And conversely, asynchronous exceptions must _not_ be
+thrown synchronously. The `UnliftIO.Exception` module has quite a few
+safeguards in place to ensure this happens; please see the "Async
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/this/this doesn't ?

File: content/tutorials/exceptions.md [hunk 6]
```diff
@@ -2,379 +2,697 @@
...
-    h2 <- openFile fp2 `onException` closeFile h1
-    restore (inner h1 h2) `finally` closeFile h2 `finally` closeFile h1
+throwString :: (MonadIO m, HasCallStack) => String -> m a
 ```
 
-The big takeaway you should get from this exercise: while masking
-prevent asynchronous exceptions from running, it does _nothing_ to
-prevent synchronous exceptions. You must still deal with those
-explicitly.
+`throwString` looks pretty similar to `error`:
```
Comment (psibi) [inline content/tutorials/exceptions.md:479]: Probably it's useful to add here that it throws an exception of type `StringException` ?

File: content/tutorials/exceptions.md [hunk 7]
```diff
@@ -2,379 +2,697 @@
...
+This uses the `Typeable` typeclass, which allows for runtime type
+analysis, which is what makes all of this magic work. Love it or hate
+it, this is at the core of the exception handling mechanism in
+Haskell.
 
-* How do we deal with asynchronous exceptions?
-* What happens if `openFile` throws an exception?
-* What happens if `closeFile` throws an exception?
-* How do we know which `IO` actions can throw an exception?
+We can also create hiearchies of exceptions. In my experience, these
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/hiearchies/hierarchies

File: content/tutorials/exceptions.md [hunk 8]
```diff
@@ -2,379 +2,697 @@
...
-impossible-to-kill thread, but make sure that we do all of our
-allocation and cleanup in a masked state so we can avoid getting
-exceptions at unexpected places. We now know for certain that only
-`IO` actions have the potential to throw an exception, and ensure that
-our code does not leak in the case of any of those actions leaking
-exceptions.
+Notice the type variable `e`. `try` will catch whichever type of
+exception you ask it to. But if you're unclear about which exception
+type you care about, the compiler will complain about ambiguous
+types. That's why, in our hiearchical exception above, I turned on
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/hiearchical/hierarchical

File: content/tutorials/exceptions.md [hunk 9]
```diff
@@ -2,379 +2,697 @@
...
+exception type. And conversely, asynchronous exceptions must _not_ be
+thrown synchronously. The `UnliftIO.Exception` module has quite a few
+safeguards in place to ensure this happens; please see the "Async
+Exception Handling in Haskell" article above for the gory details.
+
+Upshot of all of this:
+
+* If you define your own exception type for asynchronous exceptions,
+  make it a child of `SomeAsyncException`. (Note: this is a pretty
+  unusal thing to do.)
```
Comment (psibi) [inline content/tutorials/exceptions.md]: /s/unusal/unusual

**Review by psibi — CHANGES_REQUESTED** [review]
Apart from some minor comments, this looks excellent!

**Comment by snoyberg** [comment]
Thanks for the thorough review! I've made updates to address all comments. I've left two discussions unresolved. Can you check that my edits address these points sufficiently?

**Review by psibi — APPROVED** [review]

**Comment by psibi** [comment]
@snoyberg Thanks, I have resolved them. 

In the above documentation you have mentioned that `UnliftIO.Exception` is exported under `rio`. But I wasn't able to find any via hoogle:

https://www.stackage.org/lts-13.12/hoogle?q=try&package=rio

(even other hoogle queries like `SomeAsyncException`)

Am I doing something wrong ?

**Comment by snoyberg** [comment]
I've reworded slightly to state this correctly: the _contents_ of that module are reexported under the `RIO` module. Thanks for pointing that out, and for the overall review!

**Pull request merged**